The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9514398075/job/26226537963) [(plain)](https://github.com/rust-lang/rust/commit/7c8e1f1cfe753af0e62232ef8ec16c44c528190d/checks/26226537963/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.443   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.459 Collecting boolean-py==4.0
#13 0.466   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.484 Collecting chardet==5.1.0
---
#13 3.593 Building wheels for collected packages: reuse
#13 3.594   Building wheel for reuse (pyproject.toml): started
#13 3.917   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.918   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.919   Stored in directory: /tmp/pip-ephem-wheel-cache-zcvkhiqm/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.921 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.942   Attempting uninstall: setuptools
#13 3.943     Found existing installation: setuptools 59.6.0
#13 3.944     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.944     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.945     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.606 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.607 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.114 Collecting virtualenv
#13 5.165   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.325      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 24.9 MB/s eta 0:00:00
#13 5.377 Collecting platformdirs<5,>=3.9.1
#13 5.383   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.403 Collecting distlib<1,>=0.3.7
#13 5.409   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.454 Collecting filelock<4,>=3.12.2
#13 5.460   Downloading filelock-3.15.1-py3-none-any.whl (15 kB)
#13 5.460   Downloading filelock-3.15.1-py3-none-any.whl (15 kB)
#13 5.547 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.710 Successfully installed distlib-0.3.8 filelock-3.15.1 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      182208 kB
DirectMap2M:     8206336 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/d2ad293851dc8e14a61355d0358490b77efae8cb/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-d2ad293851dc8e14a61355d0358490b77efae8cb-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 25.59s
##[endgroup]
fmt check
##[error]Diff in /checkout/src/bootstrap/src/core/sanity.rs at line 128:
                     eprintln!(
                         "Consider upgrading libstdc++ or disabling the `llvm.download-ci-llvm` option."
                     );
-                    eprintln!("If you choose to upgrade libstdc++, run `x clean` or delete `build/host/libcxx-version` manually after the upgrade.");
+                    eprintln!(
+                        "If you choose to upgrade libstdc++, run `x clean` or delete `build/host/libcxx-version` manually after the upgrade."
                 }
             }
             tool::LibcxxVersion::Llvm(_) => {
             tool::LibcxxVersion::Llvm(_) => {
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/bootstrap/src/utils/exec.rs" "/checkout/src/bootstrap/src/utils/metrics.rs" "/checkout/tests/run-make/issue-51671/app.rs" "/checkout/src/bootstrap/src/utils/change_tracker/tests.rs" "/checkout/src/bootstrap/src/utils/change_tracker.rs" "/checkout/src/bootstrap/src/utils/mod.rs" "/checkout/src/bootstrap/src/utils/render_tests.rs" "/checkout/src/bootstrap/src/utils/tarball.rs" "/checkout/src/bootstrap/src/utils/job.rs" "/checkout/src/bootstrap/src/utils/cc_detect.rs" "/checkout/src/bootstrap/src/utils/channel.rs" "/checkout/src/bootstrap/src/utils/cache.rs" "/checkout/tests/run-make/compressed-debuginfo/foo.rs" "/checkout/src/bootstrap/src/utils/helpers/tests.rs" "/checkout/src/bootstrap/src/utils/dylib.rs" "/checkout/src/bootstrap/src/utils/bin_helpers.rs" "/checkout/src/bootstrap/src/utils/helpers.rs" "/checkout/src/bootstrap/src/lib.rs" "/checkout/tests/run-make/symbol-mangling-hashed/a_dylib.rs" "/checkout/tests/run-make/symbol-mangling-hashed/b_dylib.rs" "/checkout/tests/run-make/symbol-mangling-hashed/b_bin.rs" "/checkout/tests/run-make/symbol-mangling-hashed/a_rlib.rs" "/checkout/src/bootstrap/src/bin/rustc.rs" "/checkout/src/bootstrap/src/bin/rustdoc.rs" "/checkout/src/bootstrap/src/bin/sccache-plus-cl.rs" "/checkout/src/bootstrap/src/bin/main.rs" "/checkout/tests/run-make/pass-linker-flags/attribute.rs" "/checkout/tests/run-make/pass-linker-flags/empty.rs" "/checkout/tests/run-make/min-global-align/min_global_align.rs" "/checkout/src/bootstrap/src/core/sanity.rs" "/checkout/src/bootstrap/src/core/download.rs" "/checkout/tests/run-make/wasm-symbols-different-module/foo.rs" "/checkout/tests/run-make/wasm-symbols-different-module/rmake.rs" "/checkout/tests/run-make/wasm-symbols-different-module/baz.rs" "/checkout/tests/run-make/wasm-symbols-different-module/log.rs" "/checkout/tests/run-make/wasm-symbols-different-module/bar.rs" "/checkout/src/bootstrap/src/core/config/tests.rs" "/checkout/src/bootstrap/src/core/config/flags.rs" "/checkout/src/bootstrap/src/core/config/mod.rs" "/checkout/src/bootstrap/src/core/config/config.rs" "/checkout/src/bootstrap/src/core/mod.rs" "/checkout/src/bootstrap/src/core/builder.rs" "/checkout/tests/run-make/rustdoc-themes/foo.rs" "/checkout/tests/run-make/rustdoc-themes/rmake.rs" "/checkout/src/bootstrap/src/core/builder/tests.rs" "/checkout/src/bootstrap/src/core/metadata.rs" "/checkout/tests/run-make/pgo-use/main.rs" "/checkout/tests/run-make/include_bytes_deps/main.rs" "/checkout/tests/run-make/track-pgo-dep-info/main.rs" "/checkout/tests/run-make/pdb-alt-path/main.rs" "/checkout/src/bootstrap/src/core/build_steps/clippy.rs" "/checkout/src/bootstrap/src/core/build_steps/llvm.rs" "/checkout/src/bootstrap/src/core/build_steps/clean.rs" "/checkout/src/bootstrap/src/core/build_steps/test.rs" "/checkout/src/bootstrap/src/core/build_steps/mod.rs" "/checkout/src/bootstrap/src/core/build_steps/check.rs" "/checkout/src/bootstrap/src/core/build_steps/tool.rs" "/checkout/src/bootstrap/src/core/build_steps/run.rs" "/checkout/src/bootstrap/src/core/build_steps/synthetic_targets.rs" "/checkout/src/bootstrap/src/core/build_steps/format.rs" "/checkout/src/bootstrap/src/core/build_steps/vendor.rs" "/checkout/src/bootstrap/src/core/build_steps/compile.rs" "/checkout/src/bootstrap/src/core/build_steps/install.rs" "/checkout/tests/run-make/long-linker-command-lines-cmd-exe/foo.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Jun 14 10:15:05 UTC 2024
  network time: Fri, 14 Jun 2024 10:15:05 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/bootstrap/src/core/build_steps/tool.rs
+++ b/src/bootstrap/src/core/build_steps/tool.rs
@@ -823,19 +823,29 @@ impl Step for LibcxxVersionTool {
 
     fn run(self, builder: &Builder<'_>) -> LibcxxVersion {
         let out_dir = builder.out.join(self.target.to_string()).join("libcxx-version");
-        let _ = fs::remove_dir_all(&out_dir);
-        t!(fs::create_dir_all(&out_dir));
-
-        let compiler = builder.cxx(self.target).unwrap();
-        let mut cmd = Command::new(compiler);
-
         let executable = out_dir.join(exe("libcxx-version", self.target));
-        cmd.arg("-o").arg(&executable).arg(builder.src.join("src/tools/libcxx-version/main.cpp"));
 
-        builder.run_cmd(&mut cmd);
-
+        // This is a sanity-check specific step, which means it is frequently called (when using
+        // CI LLVM), and compiling `src/tools/libcxx-version/main.cpp` at the beginning of the bootstrap
+        // invocation adds a fair amount of overhead to the process (see https://github.com/rust-lang/rust/issues/126423).
+        // Therefore, we want to avoid recompiling this file unnecessarily.
         if !executable.exists() {
-            panic!("Something went wrong. {} is not present", executable.display());
+            if !out_dir.exists() {
+                t!(fs::create_dir_all(&out_dir));
+            }
+
+            let compiler = builder.cxx(self.target).unwrap();
+            let mut cmd = Command::new(compiler);
+
+            cmd.arg("-o")
+                .arg(&executable)
+                .arg(builder.src.join("src/tools/libcxx-version/main.cpp"));
+
+            builder.run_cmd(&mut cmd);
+
+            if !executable.exists() {
+                panic!("Something went wrong. {} is not present", executable.display());
+            }
         }
 
         let version_output = output(&mut Command::new(executable));

--- a/src/bootstrap/src/core/sanity.rs
+++ b/src/bootstrap/src/core/sanity.rs
@@ -128,6 +128,9 @@ pub fn check(build: &mut Build) {
                     eprintln!(
                         "Consider upgrading libstdc++ or disabling the `llvm.download-ci-llvm` option."
                     );
+                    eprintln!(
+                        "If you choose to upgrade libstdc++, run `x clean` or delete `build/host/libcxx-version` manually after the upgrade."
+                    );
                 }
             }
             tool::LibcxxVersion::Llvm(_) => {

